File: src/spotify_app_review_analyzer/deploy/vercel_build.py
"""Vercel build step: validate proxy config and write dashboard/static/js/config.js."""

from __future__ import annotations

import os
import sys
from pathlib import Path

PROJECT_ROOT = Path(__file__).resolve().parents[3]
CONFIG_PATH = PROJECT_ROOT / "dashboard" / "static" / "js" / "config.js"


def backend_url_configured() -> bool:
    return bool((os.getenv("API_BASE_URL") or os.getenv("RENDER_API_URL") or "").strip())


def resolve_api_base() -> str:
    """Use direct Render URL when set (CORS-enabled); otherwise same-origin /api + middleware."""
    raw = (os.getenv("API_BASE_URL") or os.getenv("RENDER_API_URL") or "").strip().rstrip("/")
    if raw:
        return f"{raw}/api"
    return "/api"


def main() -> None:
    api_base = resolve_api_base()
    CONFIG_PATH.parent.mkdir(parents=True, exist_ok=True)
    CONFIG_PATH.write_text(
        f"window.APP_CONFIG = {{\n  apiBase: {api_base!r},\n}};\n",
        encoding="utf-8",
    )
    mode = "direct Render URL" if api_base.startswith("http") else "same-origin /api (middleware)"
    print(f"Wrote {CONFIG_PATH} with apiBase={api_base!r} ({mode})")

    if os.getenv("VERCEL") and not backend_url_configured():
        print(
            "WARNING: API_BASE_URL is not set on Vercel. "
            "Add it under Settings → Environment Variables, then redeploy.",
            file=sys.stderr,
        )


if __name__ == "__main__":
    main()

File: scripts/vercel-build.mjs
/**
 * Vercel build: write dashboard/static/js/config.js from API_BASE_URL.
 * Uses Node so Vercel does not treat the repo as a Python/FastAPI service.
 */
import { mkdirSync, writeFileSync } from "node:fs";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";

const root = join(dirname(fileURLToPath(import.meta.url)), "..");
const raw = (process.env.API_BASE_URL || process.env.RENDER_API_URL || "")
  .trim()
  .replace(/\/$/, "");
const apiBase = raw ? `${raw}/api` : "/api";
const configPath = join(root, "dashboard", "static", "js", "config.js");

mkdirSync(dirname(configPath), { recursive: true });
writeFileSync(
  configPath,
  `window.APP_CONFIG = {\n  apiBase: ${JSON.stringify(apiBase)},\n};\n`,
  "utf8",
);

const mode = apiBase.startsWith("http") ? "direct Render URL" : "same-origin /api (middleware)";
console.log(`Wrote ${configPath} with apiBase=${JSON.stringify(apiBase)} (${mode})`);

if (process.env.VERCEL && !raw) {
  console.warn(
    "WARNING: API_BASE_URL is not set on Vercel. Add it under Settings → Environment Variables.",
  );
}
